numba/tests/test_np_functions.py

```diff
@@ -1548,6 +1551,92 @@ def test_ediff1d_exceptions(self):
         msg = "Boolean dtype is unsupported (as per NumPy)"
         assert msg in str(e.exception)
 
+    def test_extract_basic(self):
+        pyfunc = extract
+        cfunc = jit(nopython=True)(pyfunc)
+        _check = partial(self._check_output, pyfunc, cfunc)
+
+        a = np.arange(10)
+        self.rnd.shuffle(a)
+        for threshold in range(-3, 13):
+            cond = a > threshold
+            _check({'condition': cond, 'arr': a})
+
+        a = np.arange(60).reshape(4, 5, 3)
+        cond = a > 11.2
+        _check({'condition': cond, 'arr': a})
+
+        a = ((1, 2, 3), (3, 4, 5), (4, 5, 6))
+        cond = np.eye(3).flatten()
+        _check({'condition': cond, 'arr': a})
+
+        a = [1.1, 2.2, 3.3, 4.4]
+        cond = [1, 1, 0, 1]
+        _check({'condition': cond, 'arr': a})
+
+        a = np.linspace(-2, 10, 6)
+        element_pool = (True, False, np.nan, -1, -1.0, -1.2, 1, 1.0, 1.5j)
+        for cond in itertools.combinations_with_replacement(element_pool, 4):
+            _check({'condition': cond, 'arr': a})
+            _check({'condition': np.array(cond).reshape(2, 2), 'arr': a})
+
+        a = np.array([1, 2, 3])
+        cond = np.array([])
+        _check({'condition': cond, 'arr': a})
+
+        a = np.array([1, 2, 3])
+        cond = np.array([1, 0, 1, 0])  # but [1, 0, 1, 0, 1] raises
```

Comment:
I think I see what's going on... it's whether the position of the last truthy value is in bounds?! `cond = np.array([1, 0, 1, 0, 0])` should also work. I think a runtime check will be needed.

Reply:
Oh boy...

Good spot - I will update the tests and tweak the implementation

Reply:
Changed the implementation - hopefully this aligns with NumPy better.